### DISCUSSION TITLE
MACSE(?)-introduced exclamations in some alignments

### DISCUSSION THREAD

**Discussion by mchaney1** [hillerlab/TOGA#112] [Q&A] [answered]
Hello! I'm playing around with the codon alignments [found here](https://genome.senckenberg.de/download/TOGA/human_hg38_reference/MultipleCodonAlignments/), and I am finding that a small number of them contain "!" characters. It appears that these characters were introduced through some pre-processing with MACSE, but I am not sure what would be the best way to handle these characters/columns in the alignment. I can't leave them alone because most other programs will see that as an illegal character and kill the run. My temptation is to simply replace the exclamations with Ns, but I was unsure if someone else might have a more justified way of handling these characters.

Thanks in advance to anyone who might be able to provide some insight here!

**Comment by MichaelHiller** [comment, marked as answer]
Hi Morgan, 

The '!' symbol originates from MACSE v2.0, which uses it to denote frameshifts within codon alignments.
It's a bit surprising to us that MACSE finds frameshifts, as TOGA already recognizes and masks frameshifted codons.

I'll add a note tomorrow in the README. 

Michael

**Comment by kirilenkobm** [comment]
Hi @mchaney1 

As Michael already mentioned, you can safely replace them with "N".
I am a bit surprised that they arose in this dataset, I though we have already replaced them.
The updated script to produce codon alignments does not allow "!" in the output.
We will investigate, thank you for your question!

**Comment by mchaney1** [comment]
Thanks for your response on this, Michael! That sounds familiar. Given that a frameshift is what’s denoted by that character, do you think it would be best to replace ‘!’ with ‘-‘ rather than with an ‘N’?

Morgan

From: Michael Hiller ***@***.***>
Sent: Thursday, October 26, 2023 3:43 PM
To: hillerlab/TOGA ***@***.***>
Cc: Chaney, Morgan ***@***.***>; Author ***@***.***>
Subject: EXT: Re: [hillerlab/TOGA] MACSE(?)-introduced exclamations in some alignments (Discussion #112)


Hi Morgan,

The '!' symbol originates from MACSE v2.0, which uses it to denote frameshifts within codon alignments.
It's a bit surprising to us that MACSE finds frameshifts, as TOGA already recognizes and masks frameshifted codons.

I'll add a note tomorrow in the README.

Michael

—
Reply to this email directly, view it on GitHub<https://github.com/hillerlab/TOGA/discussions/112#discussioncomment-7396722>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AF3KWXK2X74A6GDU6CC7Z5LYBK4MVAVCNFSM6AAAAAA6RKD65KVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM3TGOJWG4ZDE>.
You are receiving this because you authored the thread.Message ID: ***@***.***>
CAUTION: EXTERNAL SENDER Do not click any links, open any attachments, or REPLY to the message unless you trust the sender and know the content is safe.

**Comment by kirilenkobm** [comment]
@mchaney1 
I would recommend using Ns.

I believe MACSE2.0 developers decided to use "!" marks instead of "-" to ensure the sequence lengths are divisible by 3. This is crucial as many (potential) downstream processes assume that the sequences in the codon alignments have lengths where length % 3 == 0.

**Comment by MichaelHiller** [comment]
Hi Morgan, 

I have now replaced ! by N in all codon alis. That should fix this problem once and for all. 
The files are being updated right now. 

Thx
Michael
